### ISSUE TITLE
Use RedisCluster as connection argument  for rq worker (redis 5.0.1, rq 1.15.1) 

### ISSUE THREAD

**Issue by MykolaBordakov** [rq/rq#2021]
Hi. I`m used redis version 4.6.0 and rq version 1.10.1. My code works well. 
But, after update versions to redis 5.0.1, rq 1.15.1. I starts to have an error : 
`^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/rq/worker.py", line 299, in _set_connection
    current_socket_timeout = connection.connection_pool.connection_kwargs.get("socket_timeout")
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'RedisCluster' object has no attribute 'connection_pool'`

Connection class for rq worker is RedisCluster. But, if i use redis 4.6.0 and rq version 1.15.1 i have same error. 
Can some one help me??

**Comment by yurnov** [comment]
Hi @MykolaBordakov,

could you please give some details, like code snippets and ways to reproduce the issue?

**Comment by MykolaBordakov** [comment]
Hi @yurnov 
Yes, here code exmple: 
`from redis.rediscluster import RedisCluster
from rq.worker import Worker
from rq.job import Job
from rq.queue import Queue

def main():
    queues = [env.redis_queue]
    connection = RedisCluster(host=env.redis_service, port=env.redis_port, password=env.redis_password,
                              username=env.redis_user, ssl_keyfile=env.redis_key, ssl_certfile=env.redis_cert,
                              ssl=True, ssl_cert_reqs=None)
    worker = Worker(connection=connection,
                          queues = queues ,
                          job_class=Job,
                          queue_class=Queue)
    worker.work(logging_level=env.log_level)

if name == 'main':
    main()`

Code does not works only on new rq version. ()

**Comment by peter-gy** [comment]
I experience the problem as described above.

I am establishing a connection to a `RedisCluster` as follows:

```python
from redis.cluster import RedisCluster


rc = RedisCluster(host='***', port=6379, ssl=True, decode_responses=True)
print(rc.info("server"))
print(rc.execute_command('CLUSTER', 'INFO'))
```

Outputs:

```text
{'redis_version': 7.1, 'redis_mode': 'cluster', 'arch_bits': 64, 'run_id': 0}
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:2
cluster_size:1
cluster_current_epoch:0
cluster_my_epoch:0
cluster_stats_messages_sent:0
cluster_stats_messages_received:0
total_cluster_links_buffer_limit_exceeded:0
```

Then I am trying to create an RQ `Worker` over this connection:

```python
from rq import Queue, Worker


q = Queue(connection=rc)
w = Worker(connection=rc, queues=[q])
w.work()
```

I end up with the error as follow:

```text
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 4
      1 from rq import Queue, Worker
      3 q = Queue(connection=rc)
----> 4 w = Worker(connection=rc, queues=[q])
      5 w.work()

File ~/.pyenv/versions/3.11.7/lib/python3.11/site-packages/rq/worker.py:148, in BaseWorker.__init__(self, queues, name, default_result_ttl, connection, exc_handler, exception_handlers, default_worker_ttl, maintenance_interval, job_class, queue_class, log_job_description, job_monitoring_interval, disable_default_exception_handler, prepare_for_work, serializer, work_horse_killed_handler)
    145 self.job_monitoring_interval = job_monitoring_interval
    146 self.maintenance_interval = maintenance_interval
--> 148 connection = self._set_connection(connection)
    149 self.connection = connection
    150 self.redis_server_version = None

File ~/.pyenv/versions/3.11.7/lib/python3.11/site-packages/rq/worker.py:299, in BaseWorker._set_connection(self, connection)
    297 if connection is None:
    298     connection = get_current_connection()
--> 299 current_socket_timeout = connection.connection_pool.connection_kwargs.get("socket_timeout")
    300 if current_socket_timeout is None:
    301     timeout_config = {"socket_timeout": self.connection_timeout}

AttributeError: 'RedisCluster' object has no attribute 'connection_pool'
```

This is a major blocker for our organization internally.
@yurnov, @selwin, is there any further information or reproduction I can provide?
I would be also glad to contribute a fix myself.

A more general, but related question: can you foresee any other limitations of `rq` when being used with a `RedisCluster`?
If so, what are these limitations? What do you think, what efforts would be required to address these limitations (if any)?

**Comment by yurnov** [comment]
It seems that [this PR](https://github.com/redis/redis-py/pull/2980/files) should solve this issue. It's included to redis version 5.1.0b1 and later

@MykolaBordakov @peter-gy could you please test your issue with https://github.com/redis/redis-py/releases/tag/v5.1.0b3?

**Comment by peter-gy** [comment]
@yurnov I just did a `pip install redis==5.1.0b3` but the same error is present.

```python
import redis

redis.__version__
```

outputs `'5.1.0b3'`, so I am sure the source changes got picked up.

**Comment by selwin** [comment]
Hey, so it seems like [_set_connection(connection: Redis)](https://github.com/rq/rq/blob/master/rq/worker.py#L411) does not work with RedisCluster connection.

Unfortunately I don't use Redis cluster, nor do I have access to a dev environment with Redis cluster. If you can open a pull request fixing this, I'll gladly merge this in.

Essentially what this part of the code is doing is trying to recreate a Redis connection to be used by other parts of the Worker.

**Comment by ghalimi** [comment]
@selwin We'd be happy to sponsor the project if that can help pay for a small testing Redis Cluster. It's really easy to setup on AWS. We can also help with the setup if necessary.

**Comment by selwin** [comment]
@ghalimi if you can provide a Redis Cluster that I can use to test, I can try to debug this issue sometime in the next few days.

**Comment by MykolaBordakov** [comment]
Hi @selwin 
You can cluster by docker. Here valid compose file https://github.com/bitnami/containers/blob/main/bitnami/redis-cluster/docker-compose.yml

**Comment by selwin** [comment]
Ah, in this case if Redis cluster is used, you can't use `queue.lmove`. You should fallback to `queue.lpop()` https://github.com/rq/rq/blob/2f5fecc1badb75b3e379b5d15c0b60b9c4931baa/rq/queue.py#L1364

**Comment by MykolaBordakov** [comment]
Yes. I have clean my code. Can try do PR.

**Comment by MykolaBordakov** [comment]
Hi @selwin 
Did you see my PR? 
Thank you.

**Comment by MykolaBordakov** [comment]
HI @selwin 
I accidentally closed the previous pr. I reopen it, and added your wishes.

**Comment by shyam-itspe** [comment]
```
import redis
from redis.cluster import RedisCluster, ClusterNode

from rq import Worker, Queue

startup_nodes = [
    ClusterNode(host="xxx.x.xxx.xx", port=8000),
    ClusterNode(host="xxx.x.xxx.xx", port=8001),
    ClusterNode(host="xxx.x.xxx.xx", port=8002),
]

if __name__ == '__main__':
    redis_conn = RedisCluster(startup_nodes=startup_nodes)
    queue = Queue(connection=redis_conn)
    if redis_conn.ping():
        print(redis.__version__)
        try:
            worker = Worker(queues=[queue], connection=redis_conn)
            worker.work()
        except Exception as e: 
            print(f'Error running worker: {e}')
    else:
        print("Failed to connect to redis_conn.")
```

got this as output: 
![Screenshot from 2024-09-26 11-07-25](https://github.com/user-attachments/assets/a90ac6c8-b260-4194-b50c-b817dccf0c44)

Hence it is not working with redis version 5.1.0b7 @yurnov @peter-gy

**Comment by yurnov** [comment]
> 
> Hence it is not working with redis version 5.1.0b7 @yurnov @peter-gy

yep, support of RedisCluster is not implemented yet, [here is a try](https://github.com/rq/rq/pull/2030), but that PR is not merged due to missing testing/CI part.

**Comment by lambda-science** [comment]
Following on this praying it would work one day ...
